### DISCUSSION TITLE
Can't use SPA silent check if user is signed in

### DISCUSSION THREAD

**Discussion by shedar** [zitadel/zitadel#3959] [Q&A] [answered]
Libraries like [oidc-client-ts](https://github.com/authts/oidc-client-ts) use iframe and prompt=none for silent check if a user is signed in and refresh.
But it doesn't work with zitadel. The reason is a redirect from "/oauth/v2/authorize" to "/ui/login/login" before redirecting back. The login page has "X-Frame-Options: DENY", which is understandable but breaks the iframe.
Is there a way to avoid an intermediate redirect to the login page for prompt=none to be able to use silent refresh/sign in?

**Comment by fforootd** [comment, marked as answer]
Hi @shedar 

You are right, that breaks the Iframe. 

If you want to refresh in the background you probably should use a refresh token flow instead of a silent refresh.

I think with the library you chose it should be sufficient to request the scope `offline_access` as well as allowing refresh tokens int the  zitadel settings(see screenshot below).

![image](https://user-images.githubusercontent.com/9879976/178941183-9dae3b1d-156c-405e-8fad-4da4b1347c9d.png)

**Reply by shedar** [reply]
Hi @fforootd,

Thank you for the quick reply! 

Could you please elaborate, is it a design choice to not support silent code+pkce flow, or is it something on the roadmap?

And one more question about the refresh token rotation. Since SPA is a public client, rotation [is required](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-19#section-2.2.2). Should it be implemented by a client as revocation of previous refresh tokens or is there an option to do it automatically?

**Reply by fforootd** [reply]
No problem, happy help 😁 

> Could you please elaborate, is it a design choice to not support silent code+pkce flow, or is it something on the roadmap?

It is a decision by choice from us to not support that. Since this originated as workaround in the `implicit` flow (which will be deprecated with OAuth2.1). However `silent authentication` with `prompt=none` is possible through redirects (or a pop-up) and works with code+pkce but not in an Iframe. 

> And one more question about the refresh token rotation. Since SPA is a public client, rotation [is required](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-19#section-2.2.2). Should it be implemented by a client as revocation of previous refresh tokens or is there an option to do it automatically?

ZITADEL automatically rotates the refresh token each time it is used, so there is no need to implement something specific on the client expect that it can deal with the rotate token at each usage of it.

**Comment by cocadepoma** [comment]
Hi, I know this is an old topic, but I am going through the same problem as @shedar shared.

I am using the same library `oidc-client-ts` in my React App and cannot refresh the user's tokens.

Currently, I have a 0.05-hour expiration for the access_token and id_token, and 10 days for refresh_token expiration. I am quite new on this topic and I would like to refresh the tokens before expiration but I have no idea how to perform it. Furthermore, the React examples I found are not working properly perhaps because all of them are using signinSilent.

Can you help me find a solution to refresh the tokens, please?

I am glad to help with React examples (if finally I get it working).

Thank you in advance.
